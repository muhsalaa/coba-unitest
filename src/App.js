import React, { useState } from 'react';
import axios from 'axios';
import { FaExclamationTriangle, FaCheckCircle } from 'react-icons/fa';

import { Alert, Button, Field } from './components';

function App() {
  const [credential, setCredential] = useState({ email: '', password: '' });
  const [credentialError, setCredentialError] = useState({
    email: '',
    password: '',
  });
  const [isLoading, setIsLoading] = useState(false);
  const [token, setToken] = useState('');
  const [error, setError] = useState('');

  function handleChange(e) {
    const { name, value } = e.target;
    setCredential((current) => ({ ...current, [name]: value }));
  }

  function validate() {
    const email = /\S+@\S+\.\S+/.test(credential.email)
      ? ''
      : 'Format email salahh';
    const password =
      credential.password.length < 8 ? 'Minimal password 8 karakter' : '';
    setCredentialError((current) => ({ ...current, email, password }));
    return email || password;
  }

  async function handleSubmit(e) {
    e.preventDefault();

    if (validate()) {
      return false;
    }

    setIsLoading(true);
    setError('');

    try {
      const response = await axios.post(
        'https://reqres.in/api/login',
        credential
      );

      setToken(response.data.token);
    } catch (error) {
      const message = error.response?.data?.error;
      setError(message || 'Something weird happened');
    }

    setIsLoading(false);
  }

  return (
    <div className="p-24 text-center">
      <h1 className="text-4xl mb-5 font-bold font-serif">
        Monggo login rumiyin
      </h1>
      {token && <Alert icon={FaCheckCircle}>Login Success</Alert>}
      {error && (
        <Alert color="red" icon={FaExclamationTriangle}>
          {error}
        </Alert>
      )}
      <form onSubmit={handleSubmit}>
        <Field
          name="email"
          placeholder="masukkan email"
          onChange={handleChange}
          label="Email"
          value={credential.email}
          error={credentialError.email}
        />
        <Field
          name="password"
          placeholder="masukkan password"
          type="password"
          onChange={handleChange}
          label="Password"
          value={credential.password}
          error={credentialError.password}
        />
        <Button type="submit" disabled={isLoading} block>
          Soebmit
        </Button>
      </form>
    </div>
  );
}

export default App;
