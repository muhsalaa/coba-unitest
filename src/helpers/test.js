export function hexToRGBString(hex) {
  const result = /^#?([a-f\d]{2})([a-f\d]{2})([a-f\d]{2})$/i.exec(hex);
  return result
    ? `rgb(${parseInt(result[1], 16)}, ${parseInt(result[2], 16)}, ${parseInt(
        result[3],
        16
      )})`
    : null;
}

export const findByTestAttr = (wrapper, val, elem = '') => {
  return wrapper.find(`${elem}[data-testid='${val}']`);
};
