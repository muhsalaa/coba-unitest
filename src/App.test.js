import React from 'react';
import { mount } from 'enzyme';
import toJson from 'enzyme-to-json';

import App from './App';
import { findByTestAttr } from './helpers/test';

describe('Field component', () => {
  let wrapper;
  beforeEach(() => {
    wrapper = mount(<App />);
  });

  describe('Component Renders', () => {
    it('Renders correctly and match snapshot', () => {
      expect(toJson(wrapper)).toMatchSnapshot();
    });
  });
});
