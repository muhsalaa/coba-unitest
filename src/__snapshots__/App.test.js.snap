// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Field component Component Renders Renders correctly and match snapshot 1`] = `undefined`;
